lib/Driver/Compilation.cpp

```diff
@@ -714,45 +721,81 @@ namespace driver {
     /// contents by calling \p ToolChain::constructBatchJob, then insert the
     /// new \c BatchJob into \p Batches and clear \p CurrentBatch.
     void
-    formBatchJobFromCurrentBatch(CommandSet &Batches,
-                                 llvm::SetVector<const Job *> &CurrentBatch) {
-      if (CurrentBatch.empty())
+    formBatchJobFromPartitionSet(CommandSetVector &Batches,
+                                 llvm::SetVector<const Job *> &Set) {
+      if (Set.empty())
         return;
       if (Comp.ShowJobLifecycle)
         llvm::outs() << "Forming batch job from "
-                     << CurrentBatch.size() << " constituents\n";
+                     << Set.size() << " constituents\n";
       auto const &TC = Comp.getToolChain();
-      auto J = TC.constructBatchJob(CurrentBatch.getArrayRef(), Comp);
+      auto J = TC.constructBatchJob(Set.getArrayRef(), Comp);
       if (J)
         Batches.insert(Comp.addJob(std::move(J)));
-      CurrentBatch.clear();
+      Set.clear();
     }
 
-    /// Return true iff \p Cmd can be expanded by \p CurrentBatch, meaning
-    /// that \p CurrentBatch is smaller than \p TargetBatchSize and \p Cmd
-    /// is batch-combinable with the equivalence class of \p CurrentBatch
-    /// (as represented by element 0 of \p CurrentBatch).
-    bool canExpandBatch(const Job *Cmd,
-                        llvm::SetVector<const Job *> &CurrentBatch,
-                        size_t TargetBatchSize) {
-      auto const &TC = Comp.getToolChain();
-      return (CurrentBatch.empty() ||
-              (TC.jobsAreBatchCombinable(Comp, Cmd, CurrentBatch[0]) &&
-               CurrentBatch.size() < TargetBatchSize));
+    /// Given 0 <= \p i < \p P.size(), advance i to the "next" partition,
+    /// which means filling each partition in order (distributing
+    /// any division-remainder evenly between partitions).
+    void nextPartition(size_t &i,
```

Comment:
This function probably needs a new name, since it doesn't *always* advance `i` and `Partition`.

Reply:
right!